### PR TITLE
Implement hierarchical solvers

### PR DESCRIPTION
**Pull request by jlogan-cfs**
## 11.0.0 2026-06-11

* Rust
    * !Consolidate math helper functions and use consistent array argument format
    * !Update filament, dipole, and boundary-element kernel functions and their dependencies to be generic over float type
    * !Remove Python 3.9 support
    * Improve parallel chunking by using lazy-static physical CPU count
    * Implement single-source-tree Barnes-Hut system under `hierarchical` module
    * Implement tree kernels for filaments, dipoles, and boundary-element mesh
    * Add hierarchical linear-filament and dipole benchmarks
    * Add mesh and triangle helper utilities used by hierarchical boundary-element calculations
    * Fix parallel chunk-size heuristic for single-core systems
    * Return centerline points from `filament_helix_path` when helix offset is near zero
    * Forbid unsafe Rust code at the crate level
    * Fix rustdoc lints
* Python
    * Add bindings to new hierarchical solver methods
    * Add full Python extension type stubs
    * Add hierarchical field comparison GUI example
    * Add hierarchical linear-filament self-field convergence example
    * Add tests for hierarchical solver bindings and zero-offset helix paths
    * Reorganize Python API docs by geometry type instead of field type
    * Add hierarchical result types to Python docs
    * Add hierarchical functions and diagnostics result types to public exports
    * Add `MU_0` and `inductance_matrix_axisymmetric_coaxial_rectangular_coils` to public exports
    * Correct wrapper type hints for sparse triplets and circular-to-linear mutual inductance
    * Use ty instead of pyright

### CODE REVIEW

File: src/physics/hierarchical/kernel.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    );
+
+    /// Evaluate a far-field source summary against a target summary.
+    ///
+    /// Args:
+    ///     target: Target summary value.
+    ///     source: Source summary value.
+    ///     out: Contribution value to fill.
+    ///
+    fn eval_far(
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/kernel.rs:384]: Could also call this "eval_approx" to have a more direct counterpart to "eval_exact"
(just a minor thing)

File: src/physics/hierarchical/kernel.rs [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    /// Combine child source summaries into one parent summary.
+    ///
+    /// Args:
+    ///     children: Source summary storage for the whole tree.
+    ///     child_ids: Child node indices to combine.
+    ///     out: Parent summary value to fill.
+    ///
+    /// Returns:
+    ///     Error code for the summary operation.
+    fn combine_source_summaries(
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/kernel.rs:339]: Comment from Claude

  1. combine_source_summaries has a dead parameter + a misleading doc — LOW, but the most likely to bite someone.
  The trait method (kernel.rs:316-321) takes child_ids: &[u32], and the doc says:
  - children: "Source summary storage for the whole tree"
  - child_ids: "Child node indices to combine"

  That description implies you're meant to index children[child_ids[i]]. But the evaluator (evaluator.rs:684-686) actually passes a local 2-element array of just the two child
  summaries, with child_ids as parallel global-ID metadata. And every kernel ignores child_ids — they're all written _child_ids: &[u32] and index children[0..len] locally (confirmed
  across dipole/filament/BEM). So:
  - child_ids is entirely unused dead surface area.
  - The doc is wrong in a dangerous direction: a future kernel author following it literally and writing children[child_ids[0] as usize] would index a 2-element slice with a global
  node ID (often ≫ 2) → out-of-bounds panic.

  Fix is cheap: either drop child_ids from the signature, or correct the doc to say children is the local child slice and child_ids is parallel metadata.

File: src/physics/hierarchical/kernel.rs [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+
+    /// Describe a kernel-specific error code.
+    ///
+    /// Args:
+    ///     error: Error code returned by a kernel method.
+    ///
+    /// Returns:
+    ///     Static description of the error code.
+    #[inline]
+    fn describe_error(&self, error: HierarchicalError) -> &'static str {
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/kernel.rs]: From Claude:

  2. describe_error + KernelError0/1 are unused scaffolding — LOW / dead code.
  The greps show describe_error is never called anywhere, and no kernel ever produces KernelError0 or KernelError1. So the whole kernel-specific error-description mechanism
  (kernel.rs:407-421, plus those two enum variants) is speculative. Worse: even if a kernel did return one, the Python layer Debug-formats the enum (py_hierarchical_error), so it'd
  surface as the literal "KernelError0" and describe_error's nicer string would never reach the user. Fine as forward-looking scaffolding for the GPU/FFI direction the error enum is
  clearly built for — but it should either be wired up or marked as intentionally-reserved, otherwise it reads as a loose end.

File: src/physics/hierarchical/kernel.rs [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    ///
+    /// Returns:
+    ///     Matching error value, or [`HierarchicalError::Unknown`] for unknown codes.
+    #[inline]
+    pub fn from_u32(value: u32) -> Self {
+        match value {
+            0 => Self::Ok,
+            1 => Self::EmptyInput,
+            2 => Self::LengthMismatch,
+            3 => Self::ScratchTooSmall,
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/kernel.rs]: Only cosmetic, but error-code 4 is not used -- is this resvered from something down the line?

File: src/physics/hierarchical/evaluator.rs [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+        targets,
+        moments,
+        theta,
+        out,
+        scratch,
+    )
+}
+
+#[inline]
+fn eval_validated<K, T, S, M, C, const D: usize>(
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/evaluator.rs:134]: Could use a doc-string similar to eval_scalar etc.

File: src/physics/hierarchical/evaluator.rs [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+        if left >= n_nodes || right >= n_nodes {
+            return HierarchicalError::LengthMismatch;
+        }
+    }
+
+    HierarchicalError::Ok
+}
+
+#[inline]
+fn propagate_source_summaries<K: HierarchicalKernel>(
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/evaluator.rs:671]: This stuff is really clever!

File: src/physics/hierarchical/tree.rs [hunk 7]
```diff
@@ -0,0 +1,958 @@
...
+    /// Internal node IDs grouped by tree depth from root to leaves.
+    pub(crate) internal_level_ids: &'a [u32],
+    /// CSR-style offsets into `internal_level_ids` for each internal depth.
+    pub(crate) internal_level_offsets: &'a [u32],
+    /// Maximum root-to-leaf depth.
+    pub(crate) max_depth: u32,
+}
+
+impl<T: Scalar> ClusterTree<T> {
+    /// Build a CPU-owned finalized tree using longest-axis hybrid splitting.
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/tree.rs:173]: Here is a comprehensive summary of Claude's take on tree.rs:

  What I verified — the safety-critical invariants all hold

  The evaluator's validate_source_tree_layout assumes a set of structural invariants; the whole no-panic story depends on the builders actually producing them. I traced each:

  1. Splits are always strictly interior ⇒ no empty nodes, no infinite loops. This is the single most important property. Both builders split a range at mid and recurse on [start,
  mid) / [mid, end). If mid could equal start or end, you'd get a zero-count child (violates the count != 0 invariant) and a child covering its parent's full range (infinite
  recursion). It can't:
  - median = start + count/2 is in [start+1, end-1] for all count ≥ 2 (the only case that recurses).
  - The gap-split candidate loops for i in start+1..end, so split ∈ [start+1, end-1] ⊂ (start, end).
  - accepts_hybrid_gap_split only ever narrows the choice (it can reject split and fall back to median); it never produces a boundary index.

  So every child has count ≥ 1, and since the only leaf condition is count ≤ 1 with count ≥ 1 guaranteed, every leaf has exactly leaf_count == 1 — matching the "one primitive per
  leaf" design and the tree_covers_each_input_once test.

  2. Ranges partition exactly. Children are [start, mid) and [mid, end) — disjoint, gapless, covering the parent. By induction from the root range [0, len), the leaves partition
  0..len, so each source lands in exactly one leaf. (This is what tree_covers_each_input_once asserts via the seen[] bitmap.)

  3. Node IDs and parent back-references are sound. Nodes are allocated in pop order; a parent is always allocated before its children (children frames are pushed after the parent is
  created), and each child writes node_left/right_child[parent] using the parent's already-stable ID. The Right-then-Left push order gives deterministic left-first DFS numbering.

  4. Morton Finish-frame AABB propagation runs strictly after both children. I walked the stack discipline: an internal Enter pushes Finish{node}, Enter{right}, Enter{left} (so
  Finish sits below both subtrees on the LIFO stack). Finish{node} therefore pops only after the entire left subtree, Finish{left}, the entire right subtree, and Finish{right} have
  all run — so node_aabb[node] = union(aabb[left], aabb[right]) reads finalized child boxes. Leaf children get their AABB inline in their own Enter frame, so they're ready too. The
  internal Aabb::empty() placeholder (tree.rs:620) is always overwritten before use.

  5. Morton codes stay aligned with sorted_indices. sort_indices_by_morton reorders indices and returns codes in the same final order (tree.rs:719-726), and the Morton builder then
  performs no per-node reordering — it only splits contiguous slices — so sorted_morton_codes[i] and sorted_indices[i] remain paired throughout. (The longest-axis builder, by
  contrast, re-sorts each node's slice in place; that's fine because it doesn't use Morton codes.)

  6. The level CSR enables correct bottom-up propagation. Every internal node is pushed to internal_by_depth[depth] exactly once (only on the count > 1 branch), leaves never. The
  flattened internal_level_offsets is monotonic with last offset == internal_level_ids.len(). Since the evaluator walks levels deepest→root, and a node's children are either leaves
  (summarized in the pre-pass) or internal nodes at a strictly greater depth (already processed), every combine sees finalized children. Correct.

  7. Encoding correctness. morton_code interleaves 3×21 = 63 bits (max bit position 3·20+2 = 62), staying within u64 and the nonnegative-signed range the doc claims.
  quantize_morton_coord clamps to [0,1] and rounds into [0, 2²¹−1], and short-circuits degenerate/non-finite bounds to 0 — which is exactly what makes
  morton_tree_handles_degenerate_representative_extent (all-identical points) produce a valid median-split tree instead of dividing by a zero extent.

  8. Determinism. Longest-axis uses slice::sort_by (stable) + scalar_cmp; Morton uses a (code, input_id) total-order tiebreak. Both produce deterministic trees regardless of input
  permutation with equal keys.

  Findings (all LOW / theoretical — no reachable bug)

  1. usize_to_u32 boundary can collide with the INVALID_INDEX sentinel — theoretical, unreachable. INVALID_INDEX = u32::MAX, and usize_to_u32 errors only on value > u32::MAX, so a
  node could be assigned node_id == u32::MAX == INVALID_INDEX. Reaching it needs ~4·10⁹ nodes (~2·10⁹ sources), i.e. hundreds of GB of node arrays — physically unreachable, and
  likely benign anyway since leaf-ness is decided by leaf_count, not by a child-pointer sentinel comparison. Still, the correct boundary for node-ID allocation is >= u32::MAX
  (reserve the sentinel). Worth a one-line guard or comment so it's clearly intentional. Note the up-front len > u32::MAX check (tree.rs:209) is necessary-but-not-sufficient — the
  real protection is usize_to_u32 erroring mid-build with CapacityExceeded, which is graceful.

  2. NaN / non-finite source coordinates are silently mis-placed rather than rejected — robustness, garbage-in. scalar_cmp treats NaN as Equal to everything (so a NaN-coord source
  sorts to an arbitrary stable slot), and Aabb::union skips NaN via its </> comparisons (a NaN point unioned into a box leaves the box unchanged), so a NaN source ends up outside its
  node's AABB. No panic, but the tree is spatially meaningless for that source and acceptance could misbehave. The library doesn't promise NaN handling, so this is acceptable — but
  if callers might pass un-sanitized coordinates, an explicit finite-input check (or a documented precondition) would be safer than the current silent behavior.

  3. The gap-split heuristics affect only balance/performance, not correctness. accepts_hybrid_gap_split, the dominance/min-span/min-side factors, spatial_axis_span, the count/8
  balance guard — none of these can produce an invalid split (always interior, per finding #1 above), so a mistuned heuristic degrades tree quality (skinnier trees, more work) but
  never breaks the result. Good separation: correctness doesn't ride on the heuristic. The two _splits_separated_clusters_at_* / _uses_median_for_uniform_* test pairs cover both
  branches.

  Test-coverage assessment

  Solid on the structural invariants (bijection, single-source leaves, both split branches for both builders, degenerate extents). The gaps I'd note: there's no test that randomly
  fuzzes input sizes/permutations and asserts the full invariant set (contiguous ranges, exact partition, AABB-covers-subtree, valid CSR) across many trees — the existing tests check
  specific hand-built cases. Given finding #1/#2 are unreachable/garbage-in, that's a low-priority hardening suggestion, not a defect.

  Verdict on tree.rs

  No reachable correctness bug. Both builders provably produce trees satisfying every invariant the evaluator depends on: interior-only splits (no empty nodes, no infinite
  recursion), exact range partitioning, sound parent/child wiring, correct deferred AABB propagation in the Morton path, code/index alignment, and a valid level CSR for the upward
  summary pass. The only items are a theoretical sentinel-boundary off-by-one (unreachable) and silent NaN tolerance (garbage-in) — both worth a comment or a cheap guard, neither
  worth blocking on.

Comment (mdavids-cfs) [reply src/physics/hierarchical/tree.rs:173]: The only thing I believe might be relevant is the missing NaN / non-finite source handling

File: src/physics/hierarchical/kernels/linear_filament.rs [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+pub struct LinearFilamentSummary<T: Scalar> {
+    /// `|I*dL|`-weighted origin for the net current-element source term.
+    pub origin: [T; 3],
+    /// Unit direction of the net current element after finalization.
+    pub direction: [T; 3],
+    /// Magnitude of the net current element `|sum(I*dL)|`.
+    pub magnitude: T,
+    /// Origin used for the residual magnetic dipole correction.
+    pub dipole_origin: [T; 3],
+    /// Magnetic dipole moment translated to `dipole_origin`.
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/kernels/linear_filament.rs]: Claude:

  origin and dipole_origin are always bit-for-bit identical — redundant storage and computation. I checked every site: add_source_to_summary accumulates rep_point·weight into both
  (linear_filament.rs:272-279), finalize_leaf_source_summary divides both by weight (:294-295), and combine accumulates child.{origin,dipole_origin}·weight into both (:197-208) — and
  since the children are inductively equal, the two fields can never diverge. eval_far then reads origin for the current element and dipole_origin for the dipole, but they hold the
  same value. So dipole_origin is dead weight: either collapse it into origin, or — if a distinction was intended (e.g. a current-magnitude-weighted current-element origin vs. a
  geometric dipole reference) — that distinction is unimplemented and the comment "Origin used for the residual magnetic dipole correction" is misleading. Not a bug; a cleanup.

File: src/physics/hierarchical/kernel.rs (hunk 1 above)
Comment (mdavids-cfs) [reply src/physics/hierarchical/kernel.rs:384]: On second thought (and after having read more of the code): `eval_far` is also a good and concise choice!

Comment (jlogan-cfs) [reply src/physics/hierarchical/kernel.rs:384]: Yeah, I think I'll update the other one to "eval_near" so that the distinction is more clear

File: src/physics/hierarchical/kernel.rs (hunk 2 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/kernel.rs:339]: Fixed, thanks!

File: src/physics/hierarchical/kernel.rs (hunk 1 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/kernel.rs:384]: Done

File: src/physics/hierarchical/kernel.rs (hunk 3 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/kernel.rs]: Removed, thanks!

File: src/physics/hierarchical/kernel.rs (hunk 4 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/kernel.rs]: Updated the enum layout and moved the GPU-compatibility error codes up front so that they can remain stable while the other ones change in the future.

File: src/physics/hierarchical/evaluator.rs (hunk 5 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/evaluator.rs:134]: Added docstrings for all the private functions added on this PR

File: src/physics/hierarchical/tree.rs (hunk 7 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/tree.rs:173]: 1. Fixed this - even the tiniest of edge cases will show up eventually.
2. NaN handling for coordinates is a valid concern _except_ that all output values of fields depend on the coordinates, so any NaN values will propagate to the output and produce a NaN output, _not_ a silently, subtly wrong value. I decided not to add this NaN check, because it's actually quite expensive since it requires either initializing the AABBs twice (once for the NaN check, then again for the tree construction), doing an allocation to store the AABBs and a roundtrip to RAM to retrieve them for tree construction (cache misses and RAM slowness), or doing the check during tree construction which adds a branch in the tree construction loop (slow). I'd rather just let the rare case where someone inputs an invalid value produce a NaN output.
3. Yes, the gap split is just a lightweight cluster identification pass. Any split is valid in terms of correctness, so any method of choosing the split is correct (but this one is also fast).

File: src/physics/hierarchical/kernels/linear_filament.rs (hunk 8 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/kernels/linear_filament.rs]: Fixed, thanks! That was left over from an attempt to choose a better origin for the dipole moment, and is indeed no longer needed.

File: cfsem/bindings.py [hunk 9]
```diff
@@ -25,10 +28,14 @@
     flux_density_circular_filament_cartesian as em_flux_density_circular_filament_cartesian,
 )
 from .cfsem import flux_density_dipole as em_flux_density_dipole
+from .cfsem import flux_density_dipole_hierarchical
```
Comment (mdavids-cfs) [inline cfsem/bindings.py:31]: From Claude:

  1. MEDIUM (still open) — dtype coercion gap; hierarchical functions skip the bindings.py wrapper layer. Confirmed: bindings.py:31-98 re-exports all six hierarchical functions
  directly from the compiled extension, with no em_-prefixed Python wrapper. The Rust signatures use PyReadonlyArray1<f64> / PyReadonlyArray2<f64> / PyReadonlyArray2<i64>, which
  require an exact dtype match. So a float32 loc/obs/s or an int32 triangles raises a bare TypeError from PyO3 extraction — where the sibling non-hierarchical functions silently
  up-cast via _3tup_contig / ascontiguousarray(..., dtype=float64), and the error won't name the offending argument. The contiguity half is now handled in Rust, but dtype coercion
  can only be fixed with thin Python wrappers (or by widening the Rust signatures). This is the one substantive open item in the whole binding layer.

File: src/physics/hierarchical/tests.rs [hunk 10]
```diff
@@ -0,0 +1,1980 @@
...
+            &moments,
+            &mut out,
+            &mut scratch
+        ),
+        HierarchicalError::ScratchTooSmall
+    );
+}
+
+fn run_theta_zero_matches_dense_direct_f64() {
+    let kernel = MockKernel::<f64>::new();
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/tests.rs]: Did a deep dive on test coverage with Claude, one interesting suggestion Claude made:
(the relevant section is the last paragraph, Strategy C, which would add a more localized test -- this feels optional but wanted to share that recommendation)

  What theta=0 does and does not validate

  Looking at the helper, it builds the tree, calls update_summaries, runs eval_rows at theta=0.0, and compares to eval_dense. So it genuinely exercises:
  - ✅ tree construction produces a traversable tree
  - ✅ the traversal reaches every leaf exactly once (coverage/partition)
  - ✅ eval_near matches the direct scalar physics, per primitive
  - ✅ output accumulation/scatter
  - ✅ update_summaries runs without returning an error

  But here's the structural blind spot: eval_far is never called, so nothing about the far field is checked. Specifically un-tested by theta=0:
  - ❌ accept_far ever returning true (the whole acceptance geometry)
  - ❌ eval_far — the multipole formula itself
  - ❌ combine_source_summaries — it runs inside update_summaries, but its result is never consumed (only eval_far reads summaries), so a wrong combine would pass theta=0 silently
  - ❌ the closure-ratio guard's accept path
  - ❌ the actual Barnes-Hut approximation accuracy

  So theta=0 is a strong test of the plumbing and the near kernel, and zero test of the far machinery — which is precisely the code I spent most of the review hand-verifying.

  Part 3 — How to exercise the eval_far path

  The prerequisite is the inverse of theta=0: you must make accept_far return true for at least one node. That needs theta > 0 and geometry that satisfies gap·θ > (target_diam +
  source_diam) — i.e. targets far enough from a source cluster. There are three strategies, two of which the suite already uses:

  Strategy A — a deterministic case where eval_far is provably exact → assert tight equality. This is the gold standard. linear_filament_far_cluster_uses_point_segment_source_term
  does it beautifully: two filaments at x=±1 (both +z, current 1) and a single distant target at [30, 2, 0.5]. The symmetry makes the residual dipole cancel — ½Σ(rᵢ−R)×IdLᵢ =
  [0,1,0]+[0,−1,0] = 0 — so eval_far reduces to a pure net-current-element point-segment field, which the test compares to flux_density_point_segment_scalar(...) at 1e-18 (machine
  precision). The trick: construct geometry where the neglected higher-order terms are exactly zero, so the approximation is the answer. This is the template for my recommended
  combine-identity and BEM-normalization tests — pick geometry that isolates one term.

  Strategy B — asymptotic convergence → assert relative error shrinks. When the residual terms don't vanish (a real multi-element cluster), you can't assert tight equality, so you
  push the target far (or sweep distance) and assert the error decays. boundary_element_forced_far_matches_direct_for_bent_strip_asymptotically does this — multi-triangle strip,
  forced far, assert rel < 0.1. Looser, but it's the only way to validate a genuine multipole truncation, and it's what catches a gross normalization error (e.g. the ×0.5 becoming
  ×1.0 would blow rel well past 0.1).

  Strategy C — bypass the tree, unit-test eval_far directly. Don't traverse at all: manually construct a SourceSummary (via summarize_leaf_sources + combine_source_summaries, or by
  hand), call eval_far with a known target, and compare to a hand-computed multipole value. This is the tightest localization — it tests the formula in isolation from the acceptance
  geometry, so a failure points straight at eval_far/combine rather than at the traversal. None of the current tests do this purely; it's the cleanest way to pin the ×0.5 factor and
  the moment-translation identity.

File: src/physics/hierarchical/tests.rs [hunk 11]
```diff
@@ -0,0 +1,1980 @@
...
+        HierarchicalError::Ok
+    );
+    for i in 0..bh.as_slice().len() {
+        assert!((bh[i][0] - dense[i][0]).abs() < 1e-5);
+    }
+}
+
+#[test]
+fn source_tree_theta_zero_matches_dense_direct() {
+    let kernel = MockKernel::<f64>::new();
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/tests.rs:1782]: Adding some selective rec's from Claude on potential tests:

  1. Hierarchical par-vs-serial equivalence at theta>0 with enough targets to force multi-level chunk recursion. This is the single most valuable addition. The existing par test runs
  only at theta=0 and (likely) too few targets to trigger the recursive eval_par_chunks split — so the chunk/scratch/output partitioning arithmetic and split_output_components
  no-aliasing logic I proved by hand are effectively untested in their recursive case. Add a test with ~100+ targets, theta=0.05, run par=true and par=false, and assert outputs are
  bit-identical (I proved they should be, since chunking never splits within a target). Guards the trickiest concurrency code in the PR.

File: src/physics/hierarchical/tests.rs [hunk 12]
```diff
@@ -0,0 +1,1980 @@
...
+    );
+
+    for axis in 0..3 {
+        assert!((active_only[0][axis] - with_inactive[0][axis]).abs() < 1.0e-30);
+    }
+}
+
+#[test]
+fn boundary_element_forced_far_matches_direct_for_bent_strip_asymptotically() {
+    let quad_kind = QuadratureKind::Dunavant3;
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/tests.rs:965]: Another test rec (moderate effort):

  3. Tight BEM far-summary normalization test. The ×0.5 factor I verified lives only in the summary → eval_far, which theta=0 never runs; today only the loose asymptotic
  boundary_element_forced_far_matches_direct... guards it. Add a tight unit test: for a single triangle, assert its leaf summary's current_element == triangle_current_density(...) *
  area * 0.5, or that a single far-accepted triangle's eval_far equals the direct triangle_flux_density_inner total moment. A future edit changing ×0.5 to ×1.0 would then fail loudly
  and locally instead of slipping through a convergence tolerance.

File: src/physics/hierarchical/tests.rs [hunk 13]
```diff
@@ -0,0 +1,1980 @@
...
+        &moments_b,
+        &mut summaries.node_summaries,
+    );
+    assert_eq!(err, HierarchicalError::Ok);
+    assert_eq!(summaries.node_summaries[0].moment, 12.0);
+}
+
+#[test]
+fn dipole_source_summary_centroid_tracks_moment_weights() {
+    let kernel = DipoleFluxDensityKernel::<f64>::new();
```
Comment (mdavids-cfs) [inline src/physics/hierarchical/tests.rs:1683]: Test rec (moderate effort):

  4. Residual-dipole combine identity test (filament + BEM). The moment-translation identity m_{R'} = m_R + ½(R−R')×ΣIdL and the combine associativity I traced by hand are not
  covered by theta=0 (combine never runs there). Add: build a 2–3 primitive cluster, force it into one node, and assert the combined summary's dipole_moment equals the moment
  computed directly about the parent centroid (½Σ(rᵢ−R)×IdLᵢ). The dipole kernel has an analog (dipole_source_summary_centroid_tracks_moment_weights); filament/BEM lack the
  equivalent for the translated dipole term specifically.

File: cfsem/bindings.py (hunk 9 above)
Comment (jlogan-cfs) [reply cfsem/bindings.py:31]: Yep, that's as intended

File: src/physics/hierarchical/tests.rs (hunk 10 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/tests.rs]: That test is specifically checking that the theta=0 case produces the same result as the direct calc, which uses the near-field kernel. So, it's behaving exactly as intended, and the other tests check the far-field kernel

File: src/physics/hierarchical/tests.rs (hunk 13 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/tests.rs:1683]: This example uses tree construction methods directly and never uses theta at all, which is generally only involved at evaluation time as opposed to construction time. So, no change needed

File: src/physics/hierarchical/tests.rs (hunk 12 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/tests.rs:965]: Rescaled dunavant weights to design this error out of existence

File: src/physics/hierarchical/tests.rs (hunk 11 above)
Comment (jlogan-cfs) [reply src/physics/hierarchical/tests.rs:1782]: Rescaled the dunavant weights to design this error out of existence

**Review by mdavids-cfs — APPROVED** [review]
Awesome stuff!! \o/
